collector/bcache_linux.go

```diff
@@ -0,0 +1,307 @@
+			value:           float64(ps.CacheBypassMisses),
+			metricType:      prometheus.CounterValue,
+			extraLabel:      label,
+			extraLabelValue: labelValue,
+		},
+		{
+			name:            "cache_miss_collisions_total",
+			desc:            "Instances where data insertion from cache miss raced with write (data already present).",
+			value:           float64(ps.CacheMissCollisions),
+			metricType:      prometheus.CounterValue,
+			extraLabel:      label,
+			extraLabelValue: labelValue,
+		},
+		{
+			name:            "cache_readaheads_total",
+			desc:            "Count of times readahead occurred.",
+			value:           float64(ps.CacheReadaheads),
+			metricType:      prometheus.CounterValue,
+			extraLabel:      label,
+			extraLabelValue: labelValue,
+		},
+	}
+	return metrics
+}
+
+// UpdateBcacheStats collects statistics for one bcache ID.
+func (c *bcacheCollector) updateBcacheStats(ch chan<- prometheus.Metric, s *bcache.Stats) {
+
+	const (
+		subsystem = "bcache"
+	)
+
+	var (
+		devLabel   = []string{"uuid"}
+		allMetrics []bcacheMetric
+		metrics    []bcacheMetric
+	)
+
+	allMetrics = []bcacheMetric{
+		// metrics in /sys/fs/bcache/<uuid>/
+		{
+			name:       "average_key_size_sectors",
+			desc:       "Average data per key in the btree (sectors).",
+			value:      float64(s.Bcache.AverageKeySize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_cache_size_bytes",
+			desc:       "Amount of memory currently used by the btree cache.",
+			value:      float64(s.Bcache.BtreeCacheSize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "cache_available_percent",
+			desc:       "Percentage of cache device without dirty data, useable for writeback (may contain clean cached data).",
+			value:      float64(s.Bcache.CacheAvailablePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "congested",
+			desc:       "Congestion.",
+			value:      float64(s.Bcache.Congested),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "root_usage_percent",
+			desc:       "Percentage of the root btree node in use (tree depth increases if too high).",
+			value:      float64(s.Bcache.RootUsagePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "tree_depth",
+			desc:       "Depth of the btree.",
+			value:      float64(s.Bcache.TreeDepth),
+			metricType: prometheus.GaugeValue,
+		},
+		// metrics in /sys/fs/bcache/<uuid>/internal/
+		{
+			name:       "active_journal_entries",
+			desc:       "Number of journal entries that are newer than the index.",
+			value:      float64(s.Bcache.Internal.ActiveJournalEntries),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_nodes",
+			desc:       "Total nodes in the btree.",
+			value:      float64(s.Bcache.Internal.BtreeNodes),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_read_average_duration_nanoseconds",
```

Comment:
Oh, missed that one. Our guidelines say to always export all metrics in their base unit, which is seconds for time. This makes it easier to write queries, as you don't need to make these conversions then.